devito/ir/stree/algorithms.py

```diff
@@ -150,7 +150,10 @@ def preprocess(clusters, options=None, **kwargs):
     for c in clusters:
         if c.is_halo_touch:
             hs = HaloScheme.union(e.rhs.halo_scheme for e in c.exprs)
-            queue.append(c.rebuild(halo_scheme=hs))
+            if hs.distributed_aindices:
```

Comment:
After playing with the test a bit, I'm not sure this is the right fix.

The test _does_ need a halo exchange (`u` needs to be up-to-date before getting interpolated), but what this patch is essentially: "well, if the halo_scheme is empty, drop it and don't generate anything"

However, empty halo schemes should have been dropped earlier on, to avoid clusters-level pollution. In particular, here: https://github.com/devitocodes/devito/blob/master/devito/ir/clusters/algorithms.py#L395

So, it turns out that `is_void` gives False here because the `fmapper` is non-empty while the distributed-aindices are empty. And the latter is caused by the fact that no dimensions are used to index into `u` -- just plain symbols, AFAICT

So we might have to:
* fix this somewhere else
* make the test more robust, e.g. by placing the receivers at the MPI borders

About the fix: it might need to go somewhere here https://github.com/devitocodes/devito/blob/master/devito/mpi/halo_scheme.py#L470

by the looks of it, if `f.grid.is_distributed(d)` then:

* if there's an aindex, we good (most of the cases)
* if `i[d]` is a number, we should throw a user exception, since it's illegal MPI code (can't use pure numbers to index into distributed dimensions
* if `i[d]` isn't a number (e.g., like in our test, an expression of pure symbols), then we should use a dummy dimension placeholder, capturing the fact that such dimension does require a halo exchange

Reply:
It doesn't drop it it just keeps it where it is in the list of clusters. It's dropped somewhere else probably in the mpi pass but that's a different issue.